### ISSUE TITLE
Correctly format IPv6 address in uri layer

### ISSUE THREAD

**Issue by caesar1995** [dotnet/corefx#28863]
While we closed the #28557, we were fixing the issue in `System.Net` layer, i.e., `request.RequestUri.HostNameType == UriHostNameType.IPv6 ? request.RequestUri.Host : request.RequestUri.IdnHost` (#28578), or directly appending `[]` around `uri.IdnHost` for IPv6 address (#28849).

Ideally we can fix the issue in URI layer, instead of adding workaround in the `System.Net`.

For example, this approach introduced a new issue for LLA IPv6 address: 

Uri.IdnHost will return `fe80::e077:c9a3:eeba:b8e9%18`.
Uri.Host will return `[fe80::e077:c9a3:eeba:b8e9]`.

After we change to `isIPv6Address ? uri.Host : uri.IdnHost`, the `uri.Host` will be returned if IPv6 address (no matter it's LLA or not), while the scope is mandatory for LLA.

/cc: @dotnet/ncl

**Comment by stephentoub** [comment]
This issue is marked as future, but isn't there a bug here in SocketsHttpHandler we need to fix in 2.1? That we're incorrectly dropping the scope from a link-local address?

**Comment by caesar1995** [comment]
IPv6 LLA starts with `fe80`. See [RFC 4291](https://tools.ietf.org/html/rfc4291#section-2.5.6):

```
Link-Local addresses have the following format:

   |   10     |
   |  bits    |         54 bits         |          64 bits           |
   +----------+-------------------------+----------------------------+
   |1111111010|           0             |       interface ID         |
   +----------+-------------------------+----------------------------+
```

IPv6 notation will be like: FE80::/10

**Comment by MarcoRossignoli** [comment]
>Currently it returns [fe::ee:b8e9], it should return [fe::ee:b8e9%18].

@caesar1995 ok i ask this because in you sample for LLA on issue.

**Comment by caesar1995** [comment]
> ok i ask this because in you sample for LLA on issue address start with fe.

Oops, corrected in sample to avoid confusion. Thanks!

> After this change i think...

This is a proposal which need investigation (so don't start working on it yet). If we choose to accept it, then yes we need to do those following steps.

**Comment by MarcoRossignoli** [comment]
@caesar1995 i did a pass through parsing code, let me know if i can try a PR(it this proposal success).

**Comment by rmkerr** [comment]
Hey @MarcoRossignoli if you'd like to take this issue in two segments I'm confident that we would accept a fix to part 2 here, since that behavior is clearly a bug. Part 1 requires a little more discussion, so I think splitting the fix up would be best.

**Comment by MarcoRossignoli** [comment]
@rmkerr sure! I'll work on it! Thank's for consideration!

**Comment by MarcoRossignoli** [comment]
@rmkerr i'm working on issue and when compile i get compile error 

`UriRelativeResolutionTest.cs(26,116): error CS0117: 'PlatformDetection' does not contain a definition for 'IsNetfx472OrNewer'`

i'm able to go on if remove field, any idea?

**Comment by rmkerr** [comment]
I would pull the latest version of the repo, then clean and build the whole project again. You can find more info on that here: [developer guide](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md). You probably last built the test utilities before [this PR](https://github.com/dotnet/corefx/pull/29663/files) was merged, adding the property you mentioned.

**Comment by MarcoRossignoli** [comment]
@caesar1995 @rmkerr my fault...i work on more host, i was convinced to work on updated repo, it all ok!

**Comment by kasper3** [comment]
cc @wfurt, this is also related: https://github.com/dotnet/corefx/issues/27529

**Comment by MarcoRossignoli** [comment]
@caesar1995 we merged fix to part 2 you could update issue description.

**Comment by karelz** [comment]
Thanks @MarcoRossignoli! Top post updated.

**Comment by Tratcher** [comment]
This last change is now failing in Kestrel tests and we need to make sure we have a consistent story (https://github.com/aspnet/KestrelHttpServer/issues/2637). Note these tests don't use HttpClient directly, only System.Uri and Sockets.

Do you expect the various http clients to send this scope id in the Host header? Walking the spec has an interesting history here.

https://tools.ietf.org/html/rfc7230#section-5.4
Host = uri-host [ ":" port ] ; Section 2.7.1
uri-host = <host, see [RFC3986], Section 3.2.2>
https://tools.ietf.org/html/rfc3986#section-3.2.2
host = IP-literal / IPv4address / reg-name
IP-literal = "[" ( IPv6address / IPvFuture  ) "]"
IPv6address = {... long and messy, but without scope ids ...}
   A host identified by an IPv6 literal address is represented inside
   the square brackets without a preceding version flag.  The ABNF
   provided here is a translation of the text definition of an IPv6
   literal address provided in [RFC3513].  This syntax does not support
   IPv6 scoped addressing zone identifiers.

So no, but...

Updated by https://tools.ietf.org/html/rfc6874#section-2
      IP-literal = "[" ( IPv6address / IPv6addrz / IPvFuture  ) "]"
      ZoneID = 1*( unreserved / pct-encoded )
      IPv6addrz = IPv6address "%25" ZoneID

So yes, but then https://tools.ietf.org/html/rfc6874#section-4
   An HTTP client, proxy, or other intermediary MUST remove any ZoneID
   attached to an outgoing URI, as it has only local significance at the
   sending host.

That looks like a pretty definitive No for putting them in Host headers.

Chrome, IE, and Edge do not accept scope Ids in the address bar so they won't send one. HttpClient currently does send a scope id and the request is rejected by Kestrel with a 400. That looks like a bug you'll need to address either in HttpClient or System.Uri.

**Comment by davidsh** [comment]
>Do you expect the various http clients to send this scope id in the Host header?

What does .NET Framework HttpWebRequest / HttpClient do?

**Comment by Tratcher** [comment]
.NET 4.6.1 HttpClient strips the scope id from the Host header.

**Comment by Tratcher** [comment]
With this URI change getting the host without the scopeid is a bit obnoxious.
```
internal static string GetHost(Uri requestUri)
        {
            var authority = requestUri.Authority;
            if (requestUri.HostNameType == UriHostNameType.IPv6)
            {
                // Make sure there's no % scope id. https://github.com/aspnet/KestrelHttpServer/issues/2637
                var address = IPAddress.Parse(requestUri.Host);
                address = new IPAddress(address.GetAddressBytes()); // Drop scope Id.
                authority = $"[{address}]:{requestUri.Port.ToString(CultureInfo.InvariantCulture)}";
            }
            return authority;
        }
```

**Comment by davidsh** [comment]
I think we may need to revert #29829 in the Master branch right away and reconsider how best to solve the overall problems.

**Comment by rmkerr** [comment]
I don't know if we need to revert this immediately or not, but we should definitely discuss the approach. It seems like a case where what's best from a pure URI standpoint conflicts with what's best for the HTTP stack. I'm inclined to say that the HTTP stack takes precedence, but I'd like to hear what others think.

**Comment by Tratcher** [comment]
What's best for Uri is what's best for it's consumers. What other consumers use Host, Authority, IdnHost, and what are their requirements?

**Comment by davidsh** [comment]
> I'm inclined to say that the HTTP stack takes precedence

I tend to agree with this.  I think perhaps if we really want a Uri property to return the host portion with scope-ids we might need a new property to do it. Otherwise, we risk breaking more things.

**Comment by MarcoRossignoli** [comment]
> I think perhaps if we really want a Uri property to return the host portion with scope-ids we might need a new property to do it

This could be the safest choice...even if i think it would be a "strange" property, may lead to confusion.

**Comment by Tratcher** [comment]
Hmm, I just realized this changed the output of ToString as well. .NET ToString does not include the scope, but now core does.

**Comment by davidsh** [comment]
>Hmm, I just realized this changed the output of ToString as well. .NET ToString does not include the scope, but now core does.

I don't think that's necessarily a bad thing.  The .OriginalString still has the scope-id.  In fact, we've seen an interesting interop issue using UWP apps when System.Uri needs to get converted at the lower layers to WinRT Windows.Foundation.Uri.  These uri's with scope-id that are valid System.Uri are broken when Windows.Foundation.Uri tries to re-create them (using .OriginalString).  That is a WinRT bug tracking also in GitHub corefx.

**Comment by rmkerr** [comment]
Agreed -- ideally I think the scope ID should be included in ToString. It's a meaningful part of the URI, and by removing it we change the meaning. Obviously we do have to consider app compat though, so it is worth considering what use cases we might break.

**Comment by karelz** [comment]
Triage: We may not be able to fix it at all, the goal is to decide in 5.0 timeframe.

**Comment by scalablecory** [comment]
Ran into this one today!

CC @MihaZupan

**Comment by wfurt** [comment]
I was thinking about it more and I'm not sure this is a problem other than surprise. According to documentation:

> The RFC 3490 compliant International Domain Name of the host, using Punycode as appropriate. This string, after being unescaped if necessary, is safe to use for DNS resolution.

If we put scopeID or '[]` the value will no longer be safe to use for DNS IMHO and additional work would be needed by the caller to make it passable to NameResolution API. 

I'm wondering what was the original intention for this property.
